src/s3_fetchers.rs

```diff
@@ -1,11 +1,80 @@
+use async_trait::async_trait;
 use std::str::FromStr;
 
-use aws_sdk_s3::Client;
+use aws_sdk_s3::output::{GetObjectOutput, ListObjectsV2Output};
+
+#[async_trait]
+pub trait LakeS3Client {
+    async fn get_object(
+        &self,
+        bucket: &str,
+        prefix: &str,
+    ) -> Result<GetObjectOutput, aws_sdk_s3::types::SdkError<aws_sdk_s3::error::GetObjectError>>;
+
+    async fn list_objects(
+        &self,
+        bucket: &str,
+        start_after: &str,
+    ) -> Result<
+        ListObjectsV2Output,
+        aws_sdk_s3::types::SdkError<aws_sdk_s3::error::ListObjectsV2Error>,
+    >;
+}
+
+#[derive(Clone, Debug)]
+pub struct LakeClient {
+    s3: aws_sdk_s3::Client,
+}
+
+impl LakeClient {
+    pub fn new(s3: aws_sdk_s3::Client) -> Self {
+        Self { s3 }
+    }
+}
+
+#[async_trait]
+impl LakeS3Client for LakeClient {
+    async fn get_object(
+        &self,
+        bucket: &str,
+        prefix: &str,
+    ) -> Result<GetObjectOutput, aws_sdk_s3::types::SdkError<aws_sdk_s3::error::GetObjectError>>
+    {
+        Ok(self
+            .s3
+            .get_object()
+            .bucket(bucket)
+            .key(prefix)
+            .request_payer(aws_sdk_s3::model::RequestPayer::Requester)
+            .send()
+            .await?)
+    }
+
+    async fn list_objects(
+        &self,
+        bucket: &str,
+        start_after: &str,
+    ) -> Result<
+        ListObjectsV2Output,
+        aws_sdk_s3::types::SdkError<aws_sdk_s3::error::ListObjectsV2Error>,
+    > {
+        Ok(self
+            .s3
+            .list_objects_v2()
+            .max_keys(1000) // 1000 is the default and max value for this parameter
+            .delimiter("/".to_string())
+            .start_after(start_after)
+            .request_payer(aws_sdk_s3::model::RequestPayer::Requester)
+            .bucket(bucket)
+            .send()
+            .await?)
+    }
+}
 
 /// Queries the list of the objects in the bucket, grouped by "/" delimiter.
 /// Returns the list of block heights that can be fetched
 pub(crate) async fn list_block_heights(
-    s3_client: &Client,
+    lake_client_impl: &impl LakeS3Client,
```

Comment:
```suggestion
    lake_s3_client: &impl LakeS3Client,
```

And other places like that. I think we shouldn't include `_impl` in variable names.

Reply:
done